cmd/thanos/receive.go

```diff
@@ -472,103 +473,83 @@ func setupHashring(g *run.Group,
+			statusProber.Ready()
+		}
+	}
+
+	hashringLoadFunc := func() {
+		content, err := conf.hashringsPathContent.Content()
 		if err != nil {
-			return errors.Wrap(err, "failed to initialize config watcher")
+			level.Error(logger).Log("msg", "failed to get hashring configuration content", "err", err)
+			return
+		}
+
+		// If the content is empty the default behavior is to start a single node hashring.
+		if len(content) == 0 {
+			webHandler.Hashring(receive.SingleNodeHashring(conf.endpoint))
+			level.Info(logger).Log("msg", "Empty hashring config. Set up single node hashring.")
+			hashringReadyFunc()
+			return
 		}
 
-		// Check the hashring configuration on before running the watcher.
-		if err := cw.ValidateConfig(); err != nil {
-			cw.Stop()
-			close(updates)
-			return errors.Wrap(err, "failed to validate hashring configuration file")
+		parsedConfig, err := hashringConfigLoader.ParseConfig() //nolint:ineffassign
+		if err != nil {
+			return
+		}
+
+		// nolint:ineffassign
+		hashring, err := receive.NewMultiHashring(algorithm, conf.replicationFactor, parsedConfig) //nolint:ineffassign
+		if err != nil {
+			level.Error(logger).Log("msg", "unable to create new hashring from config", "err", err)
+			return
 		}
+		webHandler.Hashring(hashring)
+		level.Info(logger).Log("msg", "Successfully set up hashring for the given hashring config.")
+		level.Debug(logger).Log("msg", "Hashring config", "config", string(content))
+		if err := dbs.SetHashringConfig(parsedConfig); err != nil {
+			level.Error(logger).Log("msg", "failed to set hashring config in MultiTSDB", "err", err)
+			return
+		}
+		hashringReadyFunc()
+	}
 
+	if conf.hashringsPathContent.Path() != "" {
 		ctx, cancel := context.WithCancel(context.Background())
 		g.Add(func() error {
-			return receive.ConfigFromWatcher(ctx, updates, cw)
-		}, func(error) {
-			cancel()
-		})
-	} else {
-		var (
-			cf  []receive.HashringConfig
-			err error
-		)
-		// The Hashrings config file content given initialize configuration from content.
-		if len(conf.hashringsFileContent) > 0 {
-			cf, err = receive.ParseConfig([]byte(conf.hashringsFileContent))
+			err := extkingpin.PathContentReloader(
+				ctx,
+				conf.hashringsPathContent,
+				logger,
+				hashringLoadFunc,
+				time.Duration(*conf.refreshInterval),
+			)
 			if err != nil {
-				close(updates)
-				return errors.Wrap(err, "failed to validate hashring configuration content")
+				return err
 			}
-		}
-
-		cancel := make(chan struct{})
-		g.Add(func() error {
-			defer close(updates)
-			updates <- cf
-			<-cancel
+			<-ctx.Done()
 			return nil
-		}, func(error) {
-			close(cancel)
+		}, func(err error) {
+			level.Error(logger).Log("msg", "hashring config reloader failed", "err", err)
+			cancel()
 		})
+	} else {
+		hashringLoadFunc()
 	}
-
-	cancel := make(chan struct{})
-	g.Add(func() error {
-
-		if enableIngestion {
-			defer close(hashringChangedChan)
```

Comment:
are we ever closing that channel now?

Reply:
Will look into fixing this. Thx for reviewing.